cvmfs/kvstore.cc

```diff
@@ -0,0 +1,121 @@
+#include <errno.h>
+#include <string.h>
+#include <assert.h>
+#include <limits.h>
+#include "kvstore.h"
+#include "util_concurrency.h"
+
+using namespace std;  // NOLINT
+
+namespace kvstore {
+
+bool MemoryKvStore::GetBuffer(const shash::Any &id, MemoryBuffer *buf) {
+  return Entries.Lookup(id, buf);
+}
+
+bool MemoryKvStore::PopBuffer(const shash::Any &id, MemoryBuffer *buf) {
+  WriteLockGuard guard(rwlock_);
+  if (!Entries.Lookup(id, buf)) return false;
+  used_bytes -= (*buf).size;
+  Entries.Forget(id);
+  return true;
+}
+
+int64_t MemoryKvStore::GetSize(const shash::Any &id) {
+  MemoryBuffer mem;
+  if (Entries.Lookup(id, &mem)) {
+    return mem.size;
+  } else {
+    return -ENOENT;
+  }
+}
+
+int64_t MemoryKvStore::GetRefcount(const shash::Any &id) {
+  MemoryBuffer mem;
+  if (Entries.Lookup(id, &mem)) {
+    return mem.refcount;
+  } else {
+    return -ENOENT;
+  }
+}
+
+bool MemoryKvStore::Ref(const shash::Any &id) {
+  WriteLockGuard guard(rwlock_);
+  MemoryBuffer mem;
+  if (Entries.Lookup(id, &mem)) {
+    assert(mem.refcount < UINT_MAX);
+    ++mem.refcount;
+    Entries.Insert(id, mem);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool MemoryKvStore::Unref(const shash::Any &id) {
+  WriteLockGuard guard(rwlock_);
+  MemoryBuffer mem;
+  if (Entries.Lookup(id, &mem) && mem.refcount > 0) {
+    --mem.refcount;
+    Entries.Insert(id, mem);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+int64_t MemoryKvStore::Read(const shash::Any &id, void *buf, size_t size, off_t offset) {
+  MemoryBuffer mem;
+  ReadLockGuard guard(rwlock_);
+  if (Entries.Lookup(id, &mem)) {
+    uint64_t copy_size = min(mem.size - offset, size);
+    memcpy(buf, static_cast<char *>(mem.address) + offset, copy_size);
+    return copy_size;
+  } else {
+    return -ENOENT;
+  }
+}
+
+bool MemoryKvStore::Commit(const shash::Any &id, const kvstore::MemoryBuffer &buf) {
+  WriteLockGuard guard(rwlock_);
+  bool existed = false;
+  MemoryBuffer mem;
+  if (Entries.Lookup(id, &mem)) {
+    Delete(id);
```

Comment:
That would overwrite the existing ref count, right?  I'm not sure if this is what we want.  We might commit something that is already in use and has refcount > 1.